benchmark/orchestrator/lib/config.py

```
@@ -0,0 +1,302 @@
+        """     
+        if benchmarks:
+            for bench in benchmarks:
+                yield BenchmarkPair(bench)        
+        else:
+            yield BenchmarkPair()
+
+    def alter_template(self, key_id, val): 
+        """
+        Recurcively search and ppdate @id config field with new value @val
+        """
+        def replace(d, key_id, val):
+            for key in list(d.keys()):
+                v = d[key]
+                if isinstance(v, dict):
+                    if isinstance(key_id, dict):
+                        if key == list(key_id.items())[0][0]:
+                            return replace(v, key_id[key], val)
+                    if replace(v, key_id, val):
+                        return True
+                else:
+                    if key == key_id:
+                        parameter_type = type(d[key])
+                        try:
+                            d[key] = parameter_type(val)
+                        except:
+                            raise InvalidBenchmarkConfig("for '{}' cannot convert val = {} to type {}".format(key,val,parameter_type))
+                        return True
+            return False
+        if not replace(self.template, key_id, val):
+            raise InvalidBenchmarkConfig("parameter %s is not supported"%key_id)
+
+    def restore_template(self):
+        """ Restore initial zstor config """
+
+        self.template = deepcopy(self._template0)
+
+    def save(self, file_name):
+        """ Save current config to file """
+
+        # prepare config for output
+        output = {'scenarios': {'scenario': self.template}}
+
+        # write scenarios to a yaml file
+        with open(file_name, 'w+') as outfile:
+            yaml.dump(output, outfile, default_flow_style=False, default_style='')
+
+    def update_deployment_config(self):
+        """ 
+        Fetch current zstor server deployment config
+                ***specific for beta2***
+        """
+        try:
```

2 things I don't like here.
- except that catch all, you don't really know what you catch, it's bad practice
- the distribution is mandatory. Which mean you can't run test with replication or even without any replication or distribution.

I would do something like
```python
self.datastor = self.template['zstor_config']['datastor']
if 'distribution' in self.datastor['pipeline']:
    distribution = self.datastor['pipeline']['distribution']
    self.data_shards_nr = int(distribution['data_shards']) + int(distribution['parity_shards'])
elif 'replication' in self.datastor['pipeline']:
    self.data_shards_nr = int(self.datastor['pipeline']['replication'])
else:
    # if nothing specified, just use 1 ztrordb
    self.data_shards_nr = 1
```

---

If I except the catch, how do I handle i `zstor_config` or `datastor` are not in the dictionary? Or it's alright is I just let it fail?

---

> if zstor_config or datastor are not in the dictionary? 

This raise a KeyError, so yo can catch that.
The problem here is that you masque the real error. so it's harder to know what was wrong.
catching specific excpetion type allow you to controle and report more detail errors

---

@zaibon , what is `replication` ?

---

Replication means we replicate the same data on multiple shards. Meaning we write the same piece of data on x amount of nodes. When reading that data we only need to read from one of those written shards, not all of them.

This used to be a separate setting, but is now implicitly defined by only defining the `data_shards` and leave the `parity_shards` property either undefined or less then 1.